File: admin/lib/aliases.js
/**
 * In-process alias resolver backed by data/coin_aliases.json.
 *
 * This is the admin (Next.js) equivalent of the Python resolver at
 * backend/services/alias/resolver.py.  Both read the same shared JSON file
 * at <project-root>/data/coin_aliases.json.
 *
 * The file is read from disk on each call — this is fine because:
 *   - It's a small file (~10-50 KB)
 *   - Only admin routes use it (not user-facing traffic)
 *   - It means edits to the JSON are picked up immediately with no restart
 */

import { readFileSync } from "fs";
import { join } from "path";

// Resolve relative to the project root.
// In Next.js, process.cwd() is the project root (where package.json is).
const ALIASES_PATH = join(process.cwd(), "data", "coin_aliases.json");

/**
 * Load and parse the alias JSON file.
 * Returns the parsed object or null on error.
 */
function loadAliasFile() {
    try {
        const raw = readFileSync(ALIASES_PATH, "utf-8");
        return JSON.parse(raw);
    } catch (err) {
        console.error(`[aliases] ✗ Failed to load ${ALIASES_PATH}: ${err.message}`);
        return null;
    }
}

/**
 * Build a flat lookup dict: lowercased alias → canonical ID.
 * Includes the canonical ID itself, all explicit aliases, and exchange symbols.
 */
function buildLookup(data) {
    const lookup = {};
    const assets = data?.assets || {};

    for (const [canonicalId, entry] of Object.entries(assets)) {
        // Canonical ID maps to itself
        lookup[canonicalId.toLowerCase()] = canonicalId;

        // All explicit aliases
        for (const alias of entry.aliases || []) {
            lookup[alias.toLowerCase()] = canonicalId;
        }

        // Exchange-specific symbols also serve as incoming aliases
        for (const sym of Object.values(entry.exchange_symbols || {})) {
            lookup[sym.toLowerCase()] = canonicalId;
        }
    }

    return lookup;
}

/**
 * Resolve any alias/symbol/name to its canonical coin ID.
 *
 * @param {string} term — case-insensitive search term
 * @returns {string|null} canonical coin ID or null
 *
 * @example
 *   resolveAlias("XBT")      // → "bitcoin"
 *   resolveAlias("btc")      // → "bitcoin"
 *   resolveAlias("Bitcoin")  // → "bitcoin"
 *   resolveAlias("unknown")  // → null
 */
export function resolveAlias(term) {
    if (!term) return null;
    const data = loadAliasFile();
    if (!data) return null;
    const lookup = buildLookup(data);
    return lookup[term.trim().toLowerCase()] ?? null;
}

/**
 * Get all canonical assets from the alias file.
 * Returns { [canonicalId]: { symbol, aliases, exchange_symbols, ... } }
 */
export function getAllAssets() {
    const data = loadAliasFile();
    return data?.assets || {};
}

/**
 * Get the total number of alias strings registered.
 */
export function getAliasStats() {
    const data = loadAliasFile();
    if (!data) return { assets: 0, aliases: 0 };
    const lookup = buildLookup(data);
    const assets = Object.keys(data.assets || {}).length;
    const aliases = Object.keys(lookup).length;
    return { assets, aliases, updated_at: data._meta?.updated_at };
}

File: realtime/normalizer/aliases.py
"""
Lightweight alias resolver for the realtime server.

Reads the shared data/coin_aliases.json file (the same one used by
the backend and admin services) and provides O(1) lookups from any
exchange symbol or name → canonical coin ID.

This is a streamlined version of backend/services/alias/resolver.py,
containing only the incoming-direction resolution needed by the
realtime pipeline (exchange symbol → canonical ID).

The alias map is loaded once at startup into an in-memory dict.
To refresh after running tools/populate_aliases/main.py, call
resolver.reload() or restart the service.
"""

import json
import logging
import os
from typing import Dict, Optional

import config

logger = logging.getLogger(__name__)


class AliasResolver:
    """
    Resolves exchange symbols/names to canonical coin IDs in O(1).

    Examples:
        resolve("XBT")     → "bitcoin"   (Kraken exchange code)
        resolve("btc")     → "bitcoin"   (standard symbol)
        resolve("Bitcoin")  → "bitcoin"   (display name)
        resolve("SOL")     → "solana"
        resolve("unknown")  → None
    """

    def __init__(self, json_path: str = config.ALIAS_JSON_PATH):
        self._json_path = os.path.abspath(json_path)
        self._lookup: Dict[str, str] = {}       # alias_lower → canonical_id
        self._symbols: Dict[str, str] = {}      # canonical_id → standard symbol
        self._load()

    def _load(self) -> None:
        """Load alias map from JSON and build in-memory lookup dict."""
        try:
            with open(self._json_path, "r") as f:
                data = json.load(f)
        except FileNotFoundError:
            logger.error(f"Alias map not found at {self._json_path}")
            logger.error("Run: python tools/populate_aliases/main.py")
            return
        except json.JSONDecodeError as e:
            logger.error(f"Invalid JSON in alias map: {e}")
            return

        assets = data.get("assets", {})

        for canonical_id, entry in assets.items():
            # Canonical ID resolves to itself
            self._lookup[canonical_id.lower()] = canonical_id

            # Standard symbol (e.g. "bitcoin" → "BTC")
            symbol = entry.get("symbol", "")
            if symbol:
                self._symbols[canonical_id] = symbol
                self._lookup[symbol.lower()] = canonical_id

            # All explicit aliases
            for alias in entry.get("aliases", []):
                self._lookup[alias.lower()] = canonical_id

            # Exchange-specific symbols (e.g. Kraken's "XBT" → "bitcoin")
            for sym in entry.get("exchange_symbols", {}).values():
                self._lookup[sym.lower()] = canonical_id

        meta = data.get("_meta", {})
        logger.info(
            f"Loaded {len(assets)} assets / {len(self._lookup)} aliases "
            f"(updated: {meta.get('updated_at', 'unknown')})"
        )

    def resolve(self, term: str) -> Optional[str]:
        """
        Resolve any alias/symbol/name to its canonical coin ID.
        Returns None if not found.
        """
        if not term:
            return None
        return self._lookup.get(term.strip().lower())

    def reload(self) -> None:
        """Reload alias map from disk without restarting the service."""
        self._lookup.clear()
        self._symbols.clear()
        self._load()

    @property
    def total_aliases(self) -> int:
        return len(self._lookup)

    @property
    def total_assets(self) -> int:
        return len(set(self._lookup.values()))
